### PR TITLE
sway: add package-option

### PR DESCRIPTION
**Pull request by SebTM**
###### Description of changes
Add package-option for sway (like in home-manager) if the user want's to have configuration (wrapper, extraSessionCommands - beside needed system-option) in home-manager.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SebTM** [comment]
```
> Explanation: The documentation contains descriptions, examples, or defaults written in DocBook. NixOS is in the process of migrating from DocBook to Markdown, and DocBook is disallowed for in-tree modules. To change your contribution to use Markdown, apply mdDoc and literalMD and use the *MD variants of option creation functions where they are available. For example:
```

I haven't included a new DocBook/mdDoc-text here.

File: nixos/modules/programs/sway.nix [hunk 1]
```diff
@@ -42,6 +42,19 @@ in {
       <https://github.com/swaywm/sway/wiki> and
       "man 5 sway" for more information'');
 
+    package = mkOption {
+      type = with types; nullOr package;
+      default = defaultSwayPackage;
+      defaultText = literalExpression "pkgs.sway";
+      description = ''
```
Comment (NickCao) [inline nixos/modules/programs/sway.nix:49]:
```suggestion
      description = lib.mkDoc ''
```

Comment (SebTM) [reply nixos/modules/programs/sway.nix]: Thanks for the help

File: nixos/modules/programs/sway.nix [hunk 2]
```diff
@@ -121,16 +134,17 @@ in {
       }
     ];
     environment = {
-      systemPackages = [ swayPackage ] ++ cfg.extraPackages;
+      systemPackages = optional (cfg.package != null) cfg.package ++ cfg.extraPackages;
       # Needed for the default wallpaper:
-      pathsToLink = [ "/share/backgrounds/sway" ];
+      pathsToLink = optional (cfg.package != null) [ "/share/backgrounds/sway" ];
```
Comment (NickCao) [inline nixos/modules/programs/sway.nix:139]:
```
nix-repl> :p lib.optional true [ "foo" ]
[ [ "foo" ] ]
```
A pretty common gotcha, should use `lib.optionals` (these namings are really confusing)

Comment (SebTM) [reply nixos/modules/programs/sway.nix]: Good catch, thanks :v:

File: nixos/modules/programs/sway.nix [hunk 3]
```diff
@@ -139,7 +153,7 @@ in {
     fonts.enableDefaultFonts = mkDefault true;
     programs.dconf.enable = mkDefault true;
     # To make a Sway session available if a display manager like SDDM is enabled:
-    services.xserver.displayManager.sessionPackages = [ swayPackage ];
+    services.xserver.displayManager.sessionPackages = optional (cfg.package != null) [ cfg.package ];
```
Comment (NickCao) [inline nixos/modules/programs/sway.nix:156]: Same issue here. You can catch these by fiddling with `nixos/tests/sway.nix`

Comment (SebTM) [reply nixos/modules/programs/sway.nix]: Fixed sorry, run the tests and should be it now ✌🏻

**Review by NickCao — APPROVED** [review]

**Comment by SebTM** [comment]
Hey @primeos,

sorry this not went well I had counted on the automation for the ping here and didn't rechecked it my apologies!

Is there something left to improve/cases to cover? As the other PR seems quite similar but covers less (with a null package)

**Comment by primeos** [comment]
@SebTM yeah, no worries, that just happens sometimes and isn't really your fault.

Tbh I'm not even sure if our current automation can ping module maintainers (it works pretty well for packages but it didn't ping me on the other PR either; IIRC it used to work somewhat reliably in the past though but that was probably directly through GitHub at that time).

> Is there something left to improve/cases to cover?

I don't think so - it LGTM (I didn't spend that much thought on it though), at least for now :)
Did you by any chance also update the Home Manager documentation? (It seems like a good idea to mention the relevant/recommended NixOS configuration there as the lack of `sway.enable = true` often seems to cause quite a few issues.)

**Comment by SebTM** [comment]
> Did you by any chance also update the Home Manager documentation?
Done - see referenced PR ✌🏻

**Comment by SebTM** [comment]
@primeos Please see the MR at home-manager for the reason - for me it's ok for now I don't know how you fee about it?

**Comment by SuperSandro2000** [comment]
Why should the package ever be null? If it shouldn't be added, shouldn't the module be disabled?

**Comment by SebTM** [comment]
Because I need the options set but install sway completely with the home-manager module so different user can have different versions / package overrides

**Comment by SuperSandro2000** [comment]
Wouldn't it be easier if the package option would be just overriden with pkgs.emptyDirectory?

**Comment by SebTM** [comment]
I don't feel this is the cleaner solution and it might fail with `${cfg.package}` where the default config is searched if package is set.

**Comment by primeos** [comment]
I also find `null` much cleaner. Then it's also "documented" via the type system while `pkgs.emptyDirectory` feels **super** hacky.

**Comment by SuperSandro2000** [comment]
We shouldn't have to carry weird edge cases in the module system just that something out of tree works. Do we have alternatives to this?

**Comment by SebTM** [comment]
I don't have one (better idea), I see home-manager as part of the ecosystem and if I rember correctly there are multiple parts also cross referencing in to docs so it's not the kind of "third party out of tree" mod I would see problematic.

Is there a concrete issue with it or is it about conventions/style?

**Pull request merged**